File: src/html_imshow.cpp

/**
 * @file html_imshow.cpp
 * @author RYO_UDON 
 * @brief HTML Image Viewer for WSL 
 * @version 0.1
 * @date 2021-10-10
 * 
 * @copyright Copyright (c) 2021
 * 
 */

#include "html_imshow.hpp"

/**
 * @brief Create HTML Header
 * 
 * @param title Title of the page
 * @param language Default language is "jp" (japanese).
 */
void HtmlImshow::CreateHeader(std::string title, std::string language="jp")
{
    //std::string header;
    _header = "<!DOCTYPE html>\n<html lang=" 
                + language 
                + "\n\t<head>\n"
                + "\t\t<title>\n"
                + "\t\t\t" + title + "\n"
                + "\t\t</title>\n"
                + "\t</head>\n"
                + "\t<body> \n";
    //return(header);
}

/**
 * @brief Create HTML Footer
 * 
 */
void HtmlImshow::CreateFooter(void)
{
    _footer = "\t</body\n</html>";
}


/**
 * @brief Create HTML Image View Space 
 * 
 * @param image_space [out] image space html code
 * @param image_path [in] path of image to show
 * @param info [in] Infotmation of image(title, width, height)
 * @param space_id priority id of html space (optional)
 */
void HtmlImshow::CreateImageViewSpace(std::string &image_space, std::string image_path, ImageInfo &info, int space_id=1)
{
    image_space = "\t\t<h" + std::to_string(space_id) + ">" + info.title + "</h" + std::to_string(space_id) + ">\n"
                    +"\t\t<img src=\"" + image_path + "\" "
                    +"width=\"" + std::to_string(info.width) + "\" "
                    +"height=\"" + std::to_string(info.height) + "\" "
                    +">\n";
                    
}


/**
 * @brief Add Image View Space in HTML file
 * 
 * @param image_path path of image to show
 * @param info Information of image (title, width, height)
 */

void HtmlImshow::AddImageViewSpace(std::string image_path, ImageInfo &info)
{
    std::string image_space;
    CreateImageViewSpace(image_space, image_path, info, _space_id);
    _image_spaces.push_back(image_space);
    //_space_id++;
}

/**
 * @brief Create HTML File to show image
 * 
 */

void HtmlImshow::OutputToHtml(void)
{
    std::ofstream outputfile(_output_file_path);
    outputfile << _header;
    for(int i=0; i < _image_spaces.size(); i++)
    {
        outputfile << _image_spaces[i];
    }
    outputfile << _footer;

    outputfile.close();

}

File: python/html_imshow.py
import os,sys


class ImageInfo:
    def __init__(self):
        self.title = ""
        self.width = 0
        self.height = 0

class HtmlImshow:
    def __init__(self, output_file_path):
        self.__output_file_path = output_file_path
        self.__header = ""
        self.__image_spaces = []
        self.__footer = ""
        self.__space_id = 1

    def CreateHeader(self, title, language="jp"):
        self.__header = "<!DOCTYPE html>\n<html lang=" \
                    + language \
                    + "\n\t<head>\n" \
                    + "\t\t<title>\n" \
                    + "\t\t\t" + title + "\n" \
                    + "\t\t</title>\n" \
                    + "\t</head>\n" \
                    + "\t<body> \n"
    
    def CreateFooter(self):
        self.__footer = "\t</body\n</html>"

    
    def CreateImageViewSpace(self, image_path_str, image_info, space_id):

        image_space_html_str = "\t\t<h" + str(space_id) + ">" + image_info.title + "</h" + str(space_id) + ">\n" \
                    +"\t\t<img src=\"" + image_path_str + "\" " \
                    +"width=\"" + str(image_info.width) + "\" " \
                    +"height=\"" +str(image_info.height) + "\" " \
                    +">\n"

        return(image_space_html_str)
    
    def AddImageViewSpace(self, image_path_str, image_info):

        image_space_html_str = self.CreateImageViewSpace(image_path_str, image_info, self.__space_id)
        self.__image_spaces.append(image_space_html_str)


    
    def OutputToHtml(self):
        with open(self.__output_file_path, "w") as f:
            f.write(self.__header)
            for html_str in self.__image_spaces:
                f.write(html_str)
            f.write(self.__footer)

        

if __name__ == '__main__':
    hi = HtmlImshow("test.html")
    hi.CreateHeader("Image View", "jp")
    hi.CreateFooter()

    info = ImageInfo()
    info.title = "Lenna"
    info.width = 256
    info.height = 256
    hi.AddImageViewSpace("../image/sample/Lenna.bmp", info)

    hi.OutputToHtml()






